### PR TITLE
Give compile_error macro examples

### LINKED ISSUE #46171: Properly document compile_error!()
https://doc.rust-lang.org/std/macro.compile_error.html

Currently says "Unconditionally causes compilation to fail with the given error message when encountered." and links to the RFC that proposed it. No examples, or otherwise explaining how it's supposed to be used.

### PR DESCRIPTION
**Pull request by Havvy**
I cannot get Rust to build at all with it complaining about GCC not being a valid C compiler or something, so letting TravisCI be my tester...

Fixes #46171

### CODE REVIEW

**Comment by Havvy** [comment]
Fixes #46171

File: src/libstd/macros.rs [hunk 1]
```diff
@@ -282,9 +282,26 @@ pub mod builtin {
...
+    ///
+    /// ```
+    /// macro_rules! give_me_foo_or_bar {
+    ///     (foo) => {};
+    ///     (bar) => {};
+    ///     ($x:ident) => {
+    ///         compile_error!("This macro only accepts `foo` or `bar`");
+    ///     }
+    /// }
+    /// ```
```
Comment (kennytm) [inline src/libstd/macros.rs]: This example is currently giving an `unused_macros` warning. Please do use the macro e.g.

```rust
give_me_foo_or_bar!(neither);
// ^ will fail at compile time with message "This macro only accepts `foo` or `bar`"
```

**Comment by kennytm** [comment]
LGTM minus Travis.

cc @rust-lang/docs

**Comment by Havvy** [comment]
Literally took your two lines there and patched them into the example.

File: src/libstd/macros.rs [hunk 2]
```diff
@@ -282,9 +282,29 @@ pub mod builtin {
 
     /// Unconditionally causes compilation to fail with the given error message when encountered.
     ///
-    /// For more information, see the [RFC].
+    /// This macro should be used when a crate uses a conditional compilation strategy to provide
+    /// better error messages for errornous conditions.
     ///
-    /// [RFC]: https://github.com/rust-lang/rfcs/blob/master/text/1695-add-error-macro.md
+    /// # Examples
+    /// Two such examples are macros and `#[cfg]` environments.
```
Comment (GuillaumeGomez) [inline src/libstd/macros.rs]: Please add an empty line before this one.

File: src/libstd/macros.rs [hunk 3]
```diff
@@ -282,9 +282,29 @@ pub mod builtin {
...
+    ///     (bar) => {};
+    ///     ($x:ident) => {
+    ///         compile_error!("This macro only accepts `foo` or `bar`");
+    ///     }
+    /// }
+    ///
+    /// give_me_foo_or_bar!(neither);
+    /// // ^ will fail at compile time with message "This macro only accepts `foo` or `bar`"
+    /// ```
+    ///
```
Comment (GuillaumeGomez) [inline src/libstd/macros.rs]: Please add a sentence in here (no need for a big one!) so we don't just have two code blocks following each other.

Comment (Havvy) [reply src/libstd/macros.rs]: Done.

**Pull request merged**